sys/xtimer/xtimer.c

```diff
@@ -210,106 +192,31 @@ void xtimer_now_timex(timex_t *out)
-{
-    unsigned irqstate = irq_disable();
-    assert(mutex != NULL && thread != NULL);
-
-    if (mutex->queue.next != MUTEX_LOCKED && mutex->queue.next != NULL) {
-        list_node_t *node = list_remove(&mutex->queue, (list_node_t *)&thread->rq_entry);
-        /* if thread was removed from the list */
-        if (node != NULL) {
-            if (mutex->queue.next == NULL) {
-                mutex->queue.next = MUTEX_LOCKED;
-            }
-            *unlocked = 1;
-
-            sched_set_status(thread, STATUS_PENDING);
-            irq_restore(irqstate);
-            sched_switch(thread->priority);
-            return;
-        }
-    }
-    *unlocked = 0;
-    irq_restore(irqstate);
-}
-
 static void _mutex_timeout(void *arg)
 {
-    /* interrupts are disabled because xtimer can spin
-     * if xtimer_set spins the callback is executed
-     * in the thread context
-     *
-     * If the xtimer spin is fixed in the future
-     * interups disable/restore can be removed
-     */
-    unsigned int irqstate = irq_disable();
-
-    mutex_thread_t *mt = arg;
-    mt->blocking = 0;
-    _mutex_remove_thread_from_waiting_queue(mt->mutex, mt->thread, &mt->dequeued);
-    irq_restore(irqstate);
+    mutex_cancel_t *mc = arg;
+    mutex_cancel(mc);
 }
 
-static int _mutex_lock_internal(mutex_t *mutex, volatile uint8_t *blocking)
+int xtimer_mutex_lock_timeout(mutex_t *mutex, uint64_t timeout)
 {
-    unsigned irqstate = irq_disable();
-
-    DEBUG("PID[%" PRIkernel_pid "]: Mutex in use.\n", thread_getpid());
-
-    if (mutex->queue.next == NULL) {
-        /* mutex is unlocked. */
-        mutex->queue.next = MUTEX_LOCKED;
-        DEBUG("PID[%" PRIkernel_pid "]: mutex_wait early out.\n",
-              thread_getpid());
-        irq_restore(irqstate);
-        return 1;
-    }
-    else if (*blocking) {
-        thread_t *me = thread_get_active();
-        DEBUG("PID[%" PRIkernel_pid "]: Adding node to mutex queue: prio: %"
-              PRIu32 "\n", thread_getpid(), (uint32_t)me->priority);
-        sched_set_status(me, STATUS_MUTEX_BLOCKED);
-        if (mutex->queue.next == MUTEX_LOCKED) {
-            mutex->queue.next = (list_node_t *)&me->rq_entry;
-            mutex->queue.next->next = NULL;
-        }
-        else {
-            thread_add_to_list(&mutex->queue, me);
-        }
-        irq_restore(irqstate);
-        thread_yield_higher();
-        /* We were woken up by scheduler. Waker removed us from queue.
-         * We have the mutex now. */
-        return 1;
-    }
-    else {
-        irq_restore(irqstate);
+    if (mutex_trylock(mutex)) {
         return 0;
     }
-}
 
-int xtimer_mutex_lock_timeout(mutex_t *mutex, uint64_t timeout)
-{
-    xtimer_t t;
-    mutex_thread_t mt = {
-        mutex, thread_get_active(), .dequeued = 0, .blocking = 1
-    };
-
-    if (timeout != 0) {
-        t.callback = _mutex_timeout;
-        t.arg = &mt;
-        xtimer_set64(&t, timeout);
+    if (timeout == 0) {
```

Comment:
Maybe it would be more sensible to do
```suggestion
    if (timeout <= XTIMER_BACKOFF) {
```

Reply:
Well, it would definitely be more sensible. If the timer spins, `mutex_cancel()` will get called prior to `mutex_lock_cancelable()`. But once `mutex_cancel()` is called, the subsequent call to `mutex_lock_cancelable()` will always fail - even if the mutex is not locked. So there is no point in spinning for some time.

On the other hand, maybe callers rely on this call taking at least `timeout` microseconds on failure. That would no longer be the case. So maybe it *does* make sense to spin?

IMO, the best would have been to have an `assert()` in place that the timeout is at least one order of magnitude bigger than `XTIMER_BACKOFF`. For shorter timeouts just using `mutex_trylock()` is likely a better idea anyway.

Reply:
True, but I would not rely on all boards (!) setting `XTIMER_BACKOFF` correctly, and since this accepted any value > 0 before, changing it would break code in strange ways.